### PR TITLE
Introduces sync endpoint

### PR DESCRIPTION
**Pull request by nickwynja**
## What's the problem you solved?

Constant network activity of connecting to peers and streaming of data can be really busy and distracting. "Sync" allows you to keep Oasis running in offline mode and sync when you want to get new stuff. Sync can be expanded upon in the future to be more useful when calling from a script.

Relates to #354. 

## What solution are you recommending?

When clicking "Connect and Sync" in settings, Oasis starts the connection scheduler, waits until you're connected to peers, waits for items to download, and then disconnects from peers.

This is best used when running Oasis in offline mode. It's basically useless in "normal" mode and will actually stop the connection scheduler and disconnect you from connected peers, which may be unexpected if you're not already offline. This could probably be improved upon.

### CODE REVIEW

**Comment by nickwynja** [comment]
I think this still needs a bit of work, actually. I'll pull in the updated service PRs when they're merged and have to add some checks to make sure I'm waiting enough time for progress to update.

**Comment by christianbundy** [comment]
I'm very excited about this btw. I added some notes in the Matrix chat, probably should have done so here instead. Let me know if you'd like to pair on this sometime!

File: src/models.js [hunk 1]
```diff
@@ -287,6 +286,18 @@ module.exports = ({ cooler, isPublic }) => {
...
+
+      if (peersList !== undefined) {
+        peersList.forEach(([address, data]) => {
+          // debug(address, data.state);
+          if (data.state === "connected") {
+            peersConnected.push([address, data]);
+          }
+        });
+      }
+      return peersConnected;
```
Comment (nickwynja) [inline src/models.js:300]: This seems really nasty and would love any recommendations on how to tighten this up.

**Comment by nickwynja** [comment]
@christianbundy I think this is in better shape now. It would be very much appreciated if you could review the progress check logic. I don't think the nested loop I'm doing is the best way possible but I'm not sure of another clean way to do it and keep track of some state. Thanks!

**Comment by christianbundy** [comment]
This looks really good, I was about to merge but wanted to check in on `indexes.current` vs `indexes.target`. Can we use 'target' everywhere? The 'target' value refers to how many bytes the database has downloaded, whereas 'current' refers to how many bytes we've indexed for the various views. Since this code only cares about downloads, I think we just want to use 'target' here, but you may be using it for a different reason?

Otherwise, this LGTM. :+1:

**Comment by nickwynja** [comment]
> The 'target' value refers to how many bytes the database has downloaded, whereas 'current' refers to how many bytes we've indexed for the various views.

Ah, I see. I noticed in some cases that `target > current` and I was assuming that meant "I know `target` number of objects exist but I have only downloaded `current`".

I'll rework a bit to only use `target` and test some more.

**Comment by christianbundy** [comment]
Sounds good, happy to merge after that! The only other [silly] thing is that the index integer refers to how many bytes are in the database rather than the number of objects. (Please disregard if this is too much information.) It's weird, but it means that you can read 4 bytes from `~/.ssb/flume/log.offset` offset N, parse those bytes as an integer (L), and then read from `N + 4` to `N + 4 + L` to get the message from the database. 

So if you have `indexes.target === 1000`, then you read from bytes 100 to 1004 and get an integer like 200, and then you read from bytes 1004 to 1204 to get the message from the database. Kinda janky, but also kinda fast. :shrug: 

More info here, if you're curious: https://github.com/flumedb/flumelog-offset#format

**Comment by nickwynja** [comment]
Interesting. I never gave it more than a passing thought but was a bit surprised that think that there was nearly a billion objects... bytes makes much more sense. 😂

**Comment by nickwynja** [comment]
Also worth mentioning that in absence of a CLI for `sync`, I've been using this `curl` command:
```
curl -f -X POST http://localhost:3000/settings/conn/sync --referer http://localhost:3000
```

**Pull request merged**